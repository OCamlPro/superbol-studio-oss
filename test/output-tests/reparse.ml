(**************************************************************************)
(*                                                                        *)
(*                        SuperBOL OSS Studio                             *)
(*                                                                        *)
(*  Copyright (c) 2022-2023 OCamlPro SAS                                  *)
(*                                                                        *)
(* All rights reserved.                                                   *)
(* This source code is licensed under the GNU Affero General Public       *)
(* License version 3 found in the LICENSE.md file in the root directory   *)
(* of this source tree.                                                   *)
(*                                                                        *)
(**************************************************************************)

open Format
open Ez_file
open FileString.OP
open Testsuite_utils

let default_parser_options =
  let exec_scanners =
    Superbol_preprocs.more [
      "SQL", Superbol_preprocs.Esql.scanner;
    ]
  in
  Cobol_parser.Options.{
    (default ~exec_scanners) with
    recovery = DisableRecovery
  }

let reparse_file ~source_format ~config filename =
  let parse ~source_format input =
    Cobol_parser.parse_simple
      ~options:default_parser_options @@
    Cobol_preproc.preprocessor
      ~options:Cobol_preproc.Options.{
          default with
          config;
          source_format
        } @@
    input
  in
  let print =
    Format.asprintf "@[%a@]@." Cobol_ptree.pp_compilation_group
  in
  match Cobol_preproc.Input.from ~filename ~f:(parse ~source_format) with
  | { result = Only Some cg; _ } -> (
      Format.printf "Parse: OK. ";
      let contents = print cg in
      match parse ~source_format:(SF SFFree) (String { contents; filename }) with
      | { result = Only Some cg'; _ } ->
        if Cobol_ptree.compare_compilation_group cg cg' = 0 then
          Format.printf "Reparse: OK."
        else
          Format.printf "Reparse: Different."
      | _ | exception _ -> Format.printf "Reparse: Failure."
  )
  | _ | exception _ -> Format.printf "Parse: Failure."

let () =
  (* Print one token per line so we can diff outputs more easily. *)
  Pretty.pp_set_margin std_formatter 3;
  let config = from_dialect Cobol_config.DIALECT.mf_strict in
  deep_iter mf_root ~glob:"*.[cC][bB][lL]"
    ~f:begin fun path ->
      printf "@[<v 1>Re-parsing `%s':@ " @@ mf_testsuite // path;
      reparse_file ~source_format:(Cobol_config.SF SFFixed) ~config
        (mf_root // path);
      printf "@]@."
    end;
  let config = Cobol_config.default in
  deep_iter ibm_root ~glob:"*.cbl"
    ~f:begin fun path ->
      printf "@[<v 1>Re-parsing `%s':@ " @@ ibm_testsuite // path;
      (* Note: we (and GnuCOBOL) don't appear to provide IBM's "EXTENDED" format
         that has lines of 252 bytes. Closest we have is xCard (with lines of
         255 bytes). *)
      reparse_file ~config ~source_format:(Cobol_config.SF SFxCard)
        (ibm_root // path);
      printf "@]@."
    end;
  let config = Cobol_config.default in
  deep_iter sql_exec_root ~glob:"*.cbl"
    ~f:begin fun path ->
      printf "@[<v 1>Re-parsing `%s':@ " @@ sql_exec_testsuite // path;
      reparse_file ~config ~source_format:(Cobol_config.Auto)
        (sql_exec_root // path);
      printf "@]@."
    end;
  let config = Cobol_config.default in
  deep_iter gixsql_root ~glob:"*.cbl"
    ~f:begin fun path ->
      printf "@[<v 1>Re-parsing `%s':@ " @@ gixsql_testsuite // path;
      reparse_file ~config ~source_format:(Cobol_config.Auto)
        (gixsql_root // path);
      printf "@]@."
    end;
;;
