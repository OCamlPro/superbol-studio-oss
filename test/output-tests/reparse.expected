Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Accept/ACCEPT.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Accept/MULTIPLIER.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Accept/Shortest.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Conditn/CONDITIONS.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Conditn/IterIf.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Indexed/DirectReadIdx.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Indexed/Seq2Index.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Indexed/SeqReadIdx.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Merge/MERGE.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Perform/MileageCount.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Perform/PERFORM1.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Perform/PERFORM3.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Perform/PERFORM4.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Relative/ReadREL.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Relative/Seq2Rel.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/ReportWriter/RepWriteA.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/ReportWriter/RepWriteB.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/ReportWriter/RepWriteFull.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/ReportWriter/RepWriteSumm.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SeqIns/seqinsert.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SeqRead/Seqread.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SeqRead/SeqreadNo88.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SeqRpt/SeqRpt.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SeqWrite/SEQWRITE.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Sort/InputSORT.CBL':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Sort/MaleSORT.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Strings/RefMod.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/Strings/UnstringFileEg.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SubProg/DateValid/DateDriver.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SubProg/DateValid/ValiDate.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SubProg/DayDiff/DayDiffDriver.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SubProg/Multiply/DriverProg.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SubProg/Multiply/Fickle.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SubProg/Multiply/MultiplyNums.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/microfocus/www.csis.ul.ie/SubProg/Multiply/Steadfast.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob1.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob10.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob11.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob12.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob13.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob14.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob15.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob16.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob17.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob2.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob3.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob4.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob5.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob6.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob7.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob8.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/ibm/ibmmainframes.com/cob9.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test01.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test02.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test03.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test04.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test05.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test06.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/sql_exec/sql_test07.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/sql_exec/sql_test08.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test09.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test10.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/sql_exec/sql_test11.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/sql_exec/sql_test12.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test13.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test14.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test15.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/sql_exec/sql_test16.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL001A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL002A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL003A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL003B.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL003C.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL004A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL004B.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL005A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL005B.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL005C.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL006A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL006B.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL006C.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL007A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL008A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL009A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL010A.cbl':
 Parse: OK. Reparse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL011A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL012A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL013A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL014A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL014B.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL014C.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL014D.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL015A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL016A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL017A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL017B.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL017C.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL017D.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL017E.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL018A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL018B.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL019A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL019B.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL020A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL021A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL021B.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL021C.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL022A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL022B.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL022C.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL023A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL023B.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL024A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL025A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL026A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL027A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL028A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL029A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL030A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL031A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL033A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL034A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL035A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL036A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL037A-MYSQL.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL037A-PGSQL.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL037A-SQLITE.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL037B-MYSQL.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL037B-PGSQL.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL037B-SQLITE.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL038A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL039A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL040A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL041A-ORACLE.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL041A.cbl':
 Parse: Failure.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL042A.cbl':
 Parse: OK. Reparse: OK.
Re-parsing `test/testsuite/sql/gixsql_test/TSQL043A.cbl':
 Parse: OK. Reparse: OK.
